cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,160 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
+      throw new InvalidArtifactException(String.format(
+          "Requested artifact version %s is older than current artifact version %s.", candidateArtifact.getVersion(),
+          currentArtifact.getVersion()));
+    }
+
+    ArtifactId newArtifactId =
+        new ArtifactId(currentArtifact.getName(), candidateArtifactVersion, currentArtifact.getScope());
+
+    Id.Artifact newArtifact = Id.Artifact.fromEntityId(Artifacts.toProtoArtifactId(appId.getParent(), newArtifactId));
+    ArtifactDetail newArtifactDetail = artifactRepository.getArtifact(newArtifact);
+    List<ApplicationConfigUpdateAction> upgradeActions = Arrays.asList(ApplicationConfigUpdateAction.UPGRADE_ARTIFACT);
+
+    try {
+      ApplicationUpdateDetails detail =
+          updateApplicationInternal(appId, appId.getParent(), appId.getApplication(), null,
+                                    currentSpec.getConfiguration(), programTerminator, newArtifactDetail,
+                                    upgradeActions, ownerAdmin.getOwner(appId), /*updateSchedules=*/false);
+      LOG.debug("Application upgrade successful. Update details: {}. Error: {}", detail.getUpdateDetails(),
+                detail.getError());
+      return new ApplicationUpdateDetails("upgrade successful.", null);
+    } catch (UnsupportedOperationException E) {
+      String errorMessage = String.format("Upgrade failed for application %s as artifact %s does not support upgrade.",
+                                          appId, newArtifact);
+      return new ApplicationUpdateDetails(new NotImplementedException(errorMessage));
+    }
+  }
+
+  // Updates an application config by applying given update actions. The app should know how to apply these actions
+  // to its config.
+  private ApplicationUpdateDetails updateApplicationInternal(ApplicationId applicationId,
+                                                             NamespaceId namespaceId,
+                                                             @Nullable String appName,
+                                                             @Nullable String appVersion,
+                                                             @Nullable String currentConfigStr,
+                                                             ProgramTerminator programTerminator,
+                                                             ArtifactDetail artifactDetail,
+                                                             List<ApplicationConfigUpdateAction> updateActions,
+                                                             @Nullable KerberosPrincipalId ownerPrincipal,
+                                                             boolean updateSchedules) throws Exception {
+    ApplicationClass appClass = Iterables.getFirst(artifactDetail.getMeta().getClasses().getApps(), null);
+    if (appClass == null) {
+      throw new InvalidArtifactException(String.format("No application class found in artifact '%s' in namespace '%s'.",
```

Comment:
InvalidArtifactException is used when there is something wrong with the artifact itself, which is not the case here. 

The ApplicationSpecification is looked up earlier, this should never happen right? If not, can throw an IllegalStateException, and add a comment that this should never happen.

Reply:
Done.